### PR TITLE
[FLINK-21090][tests] Add IT case for stop-with-savepoint and final checkpoint

### PR DESCRIPTION
**Pull request by rkhachatryan**
The PR depends on ~FLINK-23811~, ~FLINK-13973~, FLINK-23808.

### CODE REVIEW

**Review by pnowojski — COMMENTED** [review]
Thanks @rkhachatryan, I think this looks like a good starting point. I've left a couple of comments. The biggest one is that I'm struggling to follow the `checkDataFlow()` assertion.

File: flink-tests/src/test/java/org/apache/flink/test/savepoint/StopWithSavepointITCase.java
Comment (pnowojski) [reply flink-tests/src/test/java/org/apache/flink/test/savepoint/StopWithSavepointITCase.java:255]: never-mind, I think it works just fine right now.

File: flink-tests/src/test/java/org/apache/flink/test/savepoint/StopWithSavepointITCase.java
Comment (rkhachatryan) [reply flink-tests/src/test/java/org/apache/flink/test/savepoint/StopWithSavepointITCase.java]: As discussed offline, I'll address it with a lower priority (and leave a `todo` for now); I've also added `MultipleInputTransformation` which is similar (though not completely the same).

File: flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/event/TestEventQueueImpl.java
Comment (pnowojski) [inline flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/event/TestEventQueueImpl.java:28]: I think this is somewhat too complicated and hard to follow with those handlers, listeners and runnables. It took me more then 10 minutes to understand this code (and I had to open it in the IDE to click through all of the usages.

Do we need all of those extra layers of abstractions and indirections? As of now this simplified proposal https://github.com/pnowojski/flink/tree/f21090 would work just as fine. (note my other comment about squashing TestEventQueue interface with it's implementations).

File: flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/TestJobExecutor.java
Comment (pnowojski) [inline flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/TestJobExecutor.java:48]: Why do we need this indirection with first defining list of steps and only after that executing them one by one? Why can not we execute them immediately one by one?

File: flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/TestJobExecutor.java
Comment (pnowojski) [inline flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/TestJobExecutor.java:132]: nit: `sleep(1)`? In the worst case `sleep(10)`?

File: flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/command/TestCommandQueue.java
Comment (pnowojski) [inline flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/command/TestCommandQueue.java:27]: instead of adding the `TestCommandTarget` (final version is quite complicated) I would also remove this indirection with two `add` methods:
```
add(TestCommand, String operatorId, TestCommandRetention);
addToAll(TestCommand, TestCommandRetention);
```

File: flink-runtime/src/test/java/org/apache/flink/runtime/testutils/CommonTestUtils.java
Comment (pnowojski) [inline flink-runtime/src/test/java/org/apache/flink/runtime/testutils/CommonTestUtils.java:188]: I would create a separate JIRA ticket for this (if it brakes something, it would be easier to find/analyse/reference it). It sounds like it's indeed a valid fix.

However shouldn't we throw an exception if state is terminal? `waitForAllTaskRunning` suggests that this is an unexpected outcome.

**Review by pnowojski — COMMENTED** [review]
I have checked so far the test structure and left a couple of comments. I will do now a second pass over the actual tested cases.

File: flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/event/TestEventQueue.java
Comment (rkhachatryan) [reply flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/event/TestEventQueue.java:26]: I'd prefer to keep the implementations separate as they have different responsibilites.
Let me update the javadocs to make it more clear.

File: flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/event/SharedTestEventQueue.java
Comment (rkhachatryan) [reply flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/event/SharedTestEventQueue.java:24]: I'll add the javadoc here as commented above.

Though I think that exclusion in that checktyle rule is reasonable.
Even public classes not always have to have javadoc.
It's a matter of taste whether simple private classes need it.

File: flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/PartiallyFinishedSourcesITCase.java
Comment (pnowojski) [inline flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/PartiallyFinishedSourcesITCase.java:82]: Maybe let's add a flag for failover or not?

File: flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/PartiallyFinishedSourcesITCase.java
Comment (pnowojski) [inline flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/PartiallyFinishedSourcesITCase.java:78]: I think we should wait for:

1. subtask(s) to FINISH after sending finish command
2. wait for one more checkpoint to start
3. wait for one more checkpoint to complete

File: flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/validation/TestJobDataFlowValidator.java
Comment (pnowojski) [inline flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/validation/TestJobDataFlowValidator.java:101]: can you expand the error message a bit? I had a failure from this error message and I had no idea what has failed.

File: flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/TestJobExecutor.java
Comment (rkhachatryan) [reply flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/TestJobExecutor.java:48]: Yes, we can. Initially, I planned to parameterize the tests with executors, but it's unnecessary, I'll remove it.
Thanks for the suggestion.

File: flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/command/TestCommandQueue.java
Comment (rkhachatryan) [reply flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/command/TestCommandQueue.java]: We discussed offline and concluded to add explicity methods.
I removed `TestCommandTarget` and added explicit for broadcast/unicast; I also renamed retention, so now it looks like this:
```
                .sendOperatorCommand(
                        finishingOperatorID,
                        fullyFinish ? ALL_SUBTASKS : SINGLE_SUBTASK,
                        FINISH_SOURCES,
                        testJob.commandQueue)
```
Further inlining (somehting like `sendSingleSubtaskCommand`) would prevent fluent interface.
WDYT?

File: flink-tests/src/test/java/org/apache/flink/test/savepoint/StopWithSavepointITCase.java
Comment (rkhachatryan) [reply flink-tests/src/test/java/org/apache/flink/test/savepoint/StopWithSavepointITCase.java]: Comparing to the actual code without lambda, I think it does (make it more readable):
```
        for (Map.Entry<Tuple2<String, Integer>, List<TestEvent>> entry :
                eventsByOperator.entrySet()) {
            if (testJob.operatorsWithLifecycleTracking.contains(entry.getKey().f0)) {
                for (TestOperatorLifecycleValidator validator : validators) {
                    validator.validateOperatorLifecycle(
                            testJob, entry.getKey().f0, entry.getKey().f1, entry.getValue());
                }
            }
        }
```
I also don't see any difference in the stacktrace.
So if it's NIT then I'd rather just leave it as is.

File: flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/event/TestEventQueue.java
Comment (pnowojski) [reply flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/event/TestEventQueue.java:26]: What are those responsibilities? For me there is one responsibility. Provide a thread safe queue of events and listeners to process those events.

And does it really matter? You are greatly complicating the code to encapsulate a single `SharedReference<TestEventQueue> delegate;` field. For that purpose you are adding something 100 lines of code? I don't think it's worth having to jump between three files to understand what's happening in a class that could have been implemented in [50 lines of code](https://github.com/pnowojski/flink/commit/e1599aebf7ca33934c189ef99fbaeedeb4c62491#diff-12701860aed3d52137cc6baa94116d9c1952bb967ee9d784ca03891d7099158e).

(another issue) Especially that even with the current separation of concerns I'm not entirely sure if this `SharedTestEventQueue` is thread safe. For example I'm not sure if some test failures are not caused by race conditions on using it. And I would suggest to replace the `SharedReference`, two `CopyOnWriteArrayList` and `LinkedBlockingQueue` with just a single lock that would make reasoning about this class easier.

File: flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/command/TestCommandQueue.java
Comment (pnowojski) [reply flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/command/TestCommandQueue.java]: What is easier to understand by a quick glance at the code snippets below?
```
dispatch(TestCommand testCommand, TestCommandScope retention, String operatorID)`
```
or
```
dispatchToAllSubtasks(TestCommand testCommand, String operatorID);
dispatchToSingleSubtask(TestCommand testCommand, String operatorID);
```
? The first one is not self explanatory. You need to jump between many places in the code to understand how does it work. 

What is the price that we would be pay by having two methods instead of one? More methods in the interfaces? Do we really need those interfaces?`SharedTestCommandDispatcher`, `TestCommandDispatcher` and `TestCommandDispatcherImpl` should be also squashed. Again, encapsulating a single field for a class that could be implemented in 60 lines of code (including imports) is not worth adding additional 100 lines of code and having to jump through 3 additional files to understand what the code is doing.

The only two remaining interface places affected by dropping `TestCommandScope` would be:
```
    public TestJobExecutor sendOperatorCommand(
            String operatorID,
            TestCommandScope retention,
            TestCommand testCommand,
            TestCommandDispatcher commandQueue) {
        LOG.debug("sendCommand: {}", testCommand);
        commandQueue.dispatch(testCommand, retention, operatorID);
        return this;
    }

    public TestJobExecutor sendBroadcastCommand(
            TestCommandScope retention,
            TestCommand testCommand,
            TestCommandDispatcher commandQueue) {
        LOG.debug("sendCommand: {}", testCommand);
        commandQueue.broadcast(testCommand, retention);
        return this;
    }
 ```
which also should be inlined with `LOG` moved into the `TestCommandDispatcher`. Almost all of the `TestJobExecutor` could be inlined apart of methods touching `job` and `miniClusterResource`.

All in all, those changes that I highlighted above would reduce number of "jumps" in the IDE to understand how this call is working: 
```
.sendBroadcastCommand(SINGLE_SUBTASK, FAIL, testJob.commandQueue)
```
from ~6 down to a single jump in:
```
testJob.commandQueue.sendToAllSubtasksOfAllTasks(FAIL);
```

File: flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/event/TestEventQueue.java
Comment (rkhachatryan) [reply flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/event/TestEventQueue.java:26]: `SharedReference` is needed to solve serializability issues, not concurrency issues. This is one responsibility.
The other one is to notify listeners about events and manage both.

The simplification you mentioned is about removing other interfaces inside the `TestEventQueueImpl`, not about `SharedReference`.
Besides that, I don't think LoC is a very useful metric. Complexity IMO actually comes from mixing concerns, not lines of code.

I totally agree that the **internals** of `TestEventQueueImpl` should be simplified (those parts are some leftovers after changing the logic and refactoring , sorry about that). I added 61cc9c310c02dea756bc02436d7fab4654caaa1a to address that, PTAL.

File: flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/command/TestCommandQueue.java
Comment (rkhachatryan) [reply flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/command/TestCommandQueue.java]: > What is the price that we would be pay by having two methods instead of one?

> Further inlining (somehting like sendSingleSubtaskCommand) would prevent fluent interface.

Without the latter, there will be branching logic like
```
if (fullyFinish) {
  sendToAllSubtasksOfAllTasks
} else {
  sendToSingleSubtaskOfAllTasks
}
// next step
```

I'd create these methods though if you insist.

The reduction of number of jumps obiously comes from merging the two classes + interface into one. They are are split for the [same reason as](https://github.com/apache/flink/pull/16773#discussion_r688929533) SharedTestEventQueue.

File: flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/PartiallyFinishedSourcesITCase.java
Comment (rkhachatryan) [reply flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/PartiallyFinishedSourcesITCase.java:96]: I added something similar in 38d6eae2ccccd8768676472f8806fa405b042340
```
                .waitForSubtasksToFinish(finishingVertexID, fullyFinish)
                // wait for a checkpoint to complete with a finished subtask(s)
                // but skip one checkpoint that might be started before finishing
                .waitForEvent(CheckpointCompletedEvent.class, testJob.eventQueue)
                .waitForEvent(CheckpointCompletedEvent.class, testJob.eventQueue)
```
If on (2) we wait for start, then on (3) there is no guarantee that this is the checkpoint just started (maybe an earlier one, before finishing).

File: flink-runtime/src/test/java/org/apache/flink/runtime/testutils/CommonTestUtils.java
Comment (rkhachatryan) [reply flink-runtime/src/test/java/org/apache/flink/runtime/testutils/CommonTestUtils.java:198]: That's a good point.

I think we should NOT throw an exception because if the task is in a terminal state it can later become running again after a failover for example.

This is possible because there are two checks:
- graph is running &&
- task is running

which are evaluated in a loop.

I'll create a separate ticket.

File: flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/event/TestEventQueueImpl.java
Comment (rkhachatryan) [reply flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/event/TestEventQueueImpl.java:31]: As commented [above](https://github.com/apache/flink/pull/16773#discussion_r688929533)
> I totally agree that the internals of TestEventQueueImpl should be simplified (those parts are some leftovers after changing the logic and refactoring , sorry about that). I added 77ec664 to address that, PTAL.

**Comment by gaoyunhaii** [comment]
Hi Roman @rkhachatryan ,  I check the failed azure test on expired checkpoints, and it seems for aborting checkpoint properly it is used an intermediate version of https://github.com/apache/flink/pull/16800 which contains bugs, we may need to change to the merged commits or rebase to the latest master~

**Comment by rkhachatryan** [comment]
Hi @gaoyunhaii, thanks for the pointer, I've rebased the PR onto the latest master.
However, the test still fails locally.

File: flink-runtime/src/test/java/org/apache/flink/runtime/testutils/CommonTestUtils.java
Comment (rkhachatryan) [reply flink-runtime/src/test/java/org/apache/flink/runtime/testutils/CommonTestUtils.java:204]: I've created FLINK-23811 / #16845 to address this (will base this PR on that change).

File: flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/TestJobExecutor.java
Comment (rkhachatryan) [reply flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/TestJobExecutor.java]: This will save 99ms at most, or am I missing something?
(I'll change it anyways)

**Comment by gaoyunhaii** [comment]
Hi Roman @rkhachatryan , there should still be two blocker issues:

https://issues.apache.org/jira/browse/FLINK-13973
https://issues.apache.org/jira/browse/FLINK-23808

I'll try to fix these two issues as fast as possible~

File: flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/event/TestEventQueueImpl.java
Comment (rkhachatryan) [reply flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/event/TestEventQueueImpl.java:31]: Unfortunately, I had to revert some simplifications to address a race condition in 7913299066a7fdd2219634f986445d4bf4d395a1.
In particular, I added `addListener` to the EventQueue, so that it can be added before executing a command.

**Review by dawidwys — COMMENTED** [review]
I found the structure of the tests quite complicated I think they cover what we planned imo.

I will publish a few minor fixes.

**Comment by dawidwys** [comment]
How far did you get with your reviews @pnowojski @gaoyunhaii ? Are you good with the current state?

**Comment by gaoyunhaii** [comment]
@flinkbot run azure

**Comment by gaoyunhaii** [comment]
Hi @dawidwys~ the current tests are in good state from my side, we might enhance the tests to also cover cases like new source api / special state types in the future, but we could go after the current PR is merged~

**Comment by dawidwys** [comment]
Let's wait for a green Azure and merge it.

Could you please create a sub-task in https://issues.apache.org/jira/browse/FLINK-23883 listing the missing cases you are aware of?

**Comment by gaoyunhaii** [comment]
No problem, I opened the issue as https://issues.apache.org/jira/browse/FLINK-24140 ~

**Review by dawidwys — APPROVED** [review]
I'm merging the PR.

**Pull request closed without merging**
